src/v/cloud_storage/topic_manifest.cc

```diff
@@ -12,21 +12,193 @@
 #include "bytes/iobuf_ostreambuf.h"
+#include "cloud_storage/logger.h"
 #include "cloud_storage/types.h"
+#include "cluster/types.h"
 #include "hashing/xx.h"
 #include "json/writer.h"
 #include "model/timestamp.h"
 
 #include <seastar/core/coroutine.hh>
+#include <seastar/core/sstring.hh>
 
 #include <boost/lexical_cast.hpp>
+#include <boost/lexical_cast/try_lexical_convert.hpp>
 #include <fmt/ostream.h>
+#include <rapidjson/encodings.h>
+#include <rapidjson/error/en.h>
 #include <rapidjson/istreamwrapper.h>
 #include <rapidjson/ostreamwrapper.h>
+#include <rapidjson/reader.h>
 
 #include <chrono>
+#include <optional>
+#include <stdexcept>
+#include <string_view>
 
 namespace cloud_storage {
+
+struct topic_manifest_handler
+  : public rapidjson::
+      BaseReaderHandler<rapidjson::UTF8<>, topic_manifest_handler> {
+    using key_string = ss::basic_sstring<char, uint32_t, 31>;
+    bool StartObject() {
+        switch (_state) {
+        case expectManifestStart:
+            _state = expectKey;
+            return true;
+        case expectKey:
+        case expectValue:
+            return false;
+        }
+    }
+
+    bool Key(const char* str, rapidjson::SizeType length, bool /*copy*/) {
+        switch (_state) {
+        case expectKey:
+            _key = key_string(str, length);
+            _state = expectValue;
+            return true;
+        case expectManifestStart:
+        case expectValue:
+            return false;
+        }
+    }
+
+    bool String(const char* str, rapidjson::SizeType length, bool /*copy*/) {
+        std::string_view sv(str, length);
+        switch (_state) {
+        case expectValue:
+            if ("namespace" == _key) {
+                _namespace = model::ns(ss::sstring(sv));
+            } else if ("topic" == _key) {
+                _topic = model::topic(ss::sstring(sv));
+            } else if ("compression" == _key) {
+                model::compression compression_res;
+                if (boost::conversion::try_lexical_convert(
+                      sv, compression_res)) {
+                    _properties.compression = compression_res;
+                } else {
+                    return false;
+                }
+            } else if ("cleanup_policy_bitflags" == _key) {
+                model::cleanup_policy_bitflags cleanup_policy_bitflags_res;
+                if (boost::conversion::try_lexical_convert(
+                      sv, cleanup_policy_bitflags_res)) {
+                    _properties.cleanup_policy_bitflags
+                      = cleanup_policy_bitflags_res;
+                } else {
+                    return false;
+                }
+            } else if ("compaction_strategy" == _key) {
+                model::compaction_strategy compaction_strategy_res;
+                if (boost::conversion::try_lexical_convert(
+                      sv, compaction_strategy_res)) {
+                    _properties.compaction_strategy = compaction_strategy_res;
+                }
```

Comment:
What would be the proper approach for performing validation? I'm assuming throwing an exception here would make sense rather than doing something like introducing an unknown compaction strategy and then looking for it after conversion?

Reply:
I'd be tempted to set the `failbit` on the istream in operator>>, but I don't know the potential impact on the rest of the codebase. Then `return false` if `try_lexical_convert` fails. Then the `if (!Parse(...))` elsewhere will return a nice error with offset of failure.

Reply:
actually parsing fails on wrong enum value, added unit test `wrong_compaction_strategy_throws` which tries to parse otherwise perfectly fine json, it fails on wrong `compaction_strategy` enum value

Reply:
Indeed it does. I just checked the implementation of `operator>>` again and it's not what I recall. No idea what I was looking at last time.